### ISSUE TITLE
Gradient Compute Error in Embedding Layers

### ISSUE THREAD

**Issue by ruipingyin** [tensorflow/tensorflow#38130]
**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): Yes.
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): Linux Ubuntu 16.04.
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if
the issue happens on mobile device: No.
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.1.0
- Python version: 3.7.3
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: 10.1
- GPU model and memory: GTX1080 Ti / 11G

**Describe the current behavior**
I tried to conduct some operations between two tensors. The first one tensor is looked up from an Embedding Layer, and the other one is the weights of the embedding layer. When I try to compute the gradients on all trainable variables, I find that only parameters related to the first tensor are computed. The other parameters in the embedding layer can not be computed and updated. This problem only occurs in Eager Model and the results in static computation graph is correct.

**Describe the expected behavior**
The gradient should be computed and updated on the whole weights in the embedding layer.

**Standalone code to reproduce the issue** 
````
import numpy as np
import tensorflow as tf
import tensorflow.keras.backend as K
from tensorflow.keras.callbacks import TensorBoard

def loss_func(model, x, y):
    y_ = model(x)
    return tf.keras.losses.MeanSquaredError(reduction=tf.keras.losses.Reduction.AUTO)(y_true=y, y_pred=y_)

def grad(model, inputs, targets):
    with tf.GradientTape() as tape:
        loss_value = loss_func(model, inputs, targets)
    return loss_value, tape.gradient(loss_value, model.trainable_variables)

batch_size = 2
nb_item = 5
nb_hidden = 3

inputs = np.array([[1], [2]])
targets = np.random.randn(batch_size, nb_item)

input_layer = tf.keras.layers.Input((1, ), dtype=tf.int32)

embd_layer = tf.keras.layers.Embedding(nb_item, nb_hidden)

embd1 = tf.reshape(embd_layer(input_layer), [-1, nb_hidden])
all_index = tf.range(nb_item, dtype=tf.int32)
embd2 = embd_layer(all_index)

sco_mat = tf.keras.layers.Lambda(lambda x: tf.matmul(x[0], x[1], transpose_b=True))([embd1, embd2])

model = tf.keras.models.Model(inputs=input_layer, outputs=[sco_mat])

loss, grads = grad(model, inputs, targets)
print(grads[0].values.numpy())
````
Output:
> [[ 0.00897689  0.01056899  0.01004288]
>  [-0.02378326 -0.02354109 -0.01821425]]
Expect output: a tensor with shape (5, 3)

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
![bbbbbb](https://user-images.githubusercontent.com/23370871/78166038-cb13df80-747e-11ea-8b5f-5f29f20a00d9.png)
I think the Node MatMul/b should not be treated as a constant value.

**Comment by Saduf2019** [comment]
i am able to replicate this issue, please find the [gist here](https://colab.sandbox.google.com/gist/Saduf2019/ec6ec90012d0717c32267da940e06267/38130.ipynb )

**Comment by tanzhenyu** [comment]
By looking at the documentation of Lambda layer, please check the `Variables` section under keras/layers/core -> class Lambda, i.e.:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/core.py#L770-L797

**Comment by ruipingyin** [comment]
> By looking at the documentation of Lambda layer, doesn't seem stateful lambda layer is encouraged. Please check the `Variables` section under keras/layers/core -> class Lambda, i.e.:
> https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/core.py#L770-L797

I have used a subclass layer to instead of the lambda layer as follows. But it seems that the results is still incorrect.

```
import numpy as np
import tensorflow as tf
import tensorflow.keras.backend as K

def loss_func(model, x, y):
    y_ = model(x)
    return tf.keras.losses.MeanSquaredError(reduction=tf.keras.losses.Reduction.AUTO)(y_true=y, y_pred=y_)

def grad(model, inputs, targets):
    with tf.GradientTape() as tape:
        loss_value = loss_func(model, inputs, targets)
    return loss_value, tape.gradient(loss_value, model.trainable_variables)

class ScoMatLayer(tf.keras.layers.Layer):
    def __init__(self, embd_layer):
        super(ScoMatLayer, self).__init__()
        all_index = tf.range(nb_item, dtype=tf.int32)
        self.embd2 = embd_layer(all_index)
    def call(self, inputs):
        return tf.matmul(inputs, self.embd2, transpose_b=True)

batch_size = 2
nb_item = 5
nb_hidden = 3

inputs = np.array([[1], [2]])
targets = np.random.randn(batch_size, nb_item)

input_layer = tf.keras.layers.Input((1, ), dtype=tf.int32)

embd_layer = tf.keras.layers.Embedding(nb_item, nb_hidden)

embd1 = tf.reshape(embd_layer(input_layer), [-1, nb_hidden])

scl = ScoMatLayer(embd_layer)

sco_mat = scl(embd1)

model = tf.keras.models.Model(inputs=input_layer, outputs=[sco_mat])

loss, grads = grad(model, inputs, targets)
print(grads[0].values.numpy())
```

> Output:
> [[-0.01078839 -0.00246409 -0.01437871]
>  [-0.01240485 -0.01926923  0.00199004]]

As you can see in this code, I think the ScoMatLayer is a stateless layer.

**Comment by tanzhenyu** [comment]
Use tf.identity(weights)?

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tf Nightly 2.6.0-dev20210524, please find the gist [here](https://colab.research.google.com/drive/1OnKnoDXOcZJSpmViPbB4ZjWWlz77PO10?resourcekey=0-Vo6nWmCqjSLuukP5d-5VrQ#scrollTo=jIpDRl8pd3tN). Thanks!

**Comment by chunduriv** [comment]
@ruipingyin, Sorry for late response.

Did you tried as suggested by @tanzhenyu to use tf.identity(weights)? Also, Many bugs have been fixed in the latest version. Can you please execute your code using `TFv2.7` and let us know if the issue still persists? Thanks!
